android/src/main/java/com/oney/WebRTCModule/PeerConnectionObserver.java

```diff
@@ -452,9 +362,154 @@ public void onSignalingChange(PeerConnection.SignalingState signalingState) {
         webRTCModule.sendEvent("peerConnectionSignalingStateChanged", params);
     }
 
+    /**
+     * Triggered when a new track is signaled by the remote peer, as a result of
+     * setRemoteDescription.
+     */
     @Override
     public void onAddTrack(final RtpReceiver receiver, final MediaStream[] mediaStreams) {
         Log.d(TAG, "onAddTrack");
+
+        // Get The timestamp so that we can reorder the transceivers on the JS layer
+        Long tsLong = System.currentTimeMillis()/1000;
+        String ts = tsLong.toString();
+
+        MediaStreamTrack track = receiver.track();
+
+        if (track == null) {
+            return;
+        }
+        if(track.kind().equals(MediaStreamTrack.VIDEO_TRACK_KIND)){
+            videoTrackAdapters.addAdapter((VideoTrack) track);
```

Comment:
Remote tracks are disposed when the RtpReceiver is disposed. We need to make sure we have removed the adapter by then or we'll be holding on to a released remote track.

Reply:
The spec does not say anything about destroying remote tracks after receiving an `onRemoveTrack` event. It mentions muting them if they are not already. So, I think it is just safest to destroy the adapters at the end of the peer connection session

Reply:
So there will be a bunch of zombie MediaStreamTrack objects? That sounds weird. Specially since transceivers can be reused, right? Can you please double check with a fiddle in Chrome?

Reply:
Have a look at [this](https://jsfiddle.net/pub2kjmf/). I tested it out on Chrome and it seems that the track still exists after a call to removeTrack. You will see that `onmute` but `onended` is not called. Also, looking at the chromium webrtc source: 
```cpp
// Called when signaling indicates that media will no longer be received on a
// track.
// With Plan B semantics, the given receiver will have been removed from the
// PeerConnection and the track muted.
// With Unified Plan semantics, the receiver will remain but the transceiver
// will have changed direction to either sendonly or inactive.
// https://w3c.github.io/webrtc-pc/#process-remote-track-removal
// TODO(hbos,deadbeef): Make pure virtual when all subclasses implement it.
virtual void OnRemoveTrack(
  rtc::scoped_refptr<RtpReceiverInterface> receiver) {}
```

It does not say anything about destroying the track.

Reply:
One leas thing to worry about I guess!